### PR TITLE
Emit diagnostic for any referenced operation with a container route prefix

### PR DESCRIPTION
**Pull request by daviwil**
This change fixes Azure/typespec-azure#3333 by adding a validation step in the `@typespec/http` library which checks whether any referenced operation lives inside of an interface or namespace with a `@route` prefix defined.  

This check is intended to help a spec author avoid the case where an operation loses the prefix of its referenced operation because it is used outside of the original container.  The issue was originally raised in issue #1466.

### CODE REVIEW

File: packages/http/src/validate.ts [hunk 1]
```diff
@@ -1,7 +1,66 @@
-import { Program } from "@typespec/compiler";
+import { Operation, Program, navigateProgram } from "@typespec/compiler";
 import { reportDiagnostic } from "./lib.js";
 import { getAllHttpServices } from "./operations.js";
-import { isSharedRoute } from "./route.js";
+import { getRoutePath, isSharedRoute } from "./route.js";
+import { OperationContainer } from "./types.js";
+
+function checkOperationReferenceContainerRoutes(program: Program) {
```
Comment (tjprescott) [inline packages/http/src/validate.ts:7]: I believe this should be a linter rule warning. `$onValidate` is typically reserved for errors.

**Review by tjprescott — CHANGES_REQUESTED** [review]

File: packages/http/src/validate.ts (hunk 1 above)
Comment (tjprescott) [reply packages/http/src/validate.ts:7]: @catalinaperalta can you confirm?

Comment (daviwil) [reply packages/http/src/validate.ts:7]: OK, I've looked at @catalinaperalta's changes to migrate the rules in `typespec-azure` and it makes sense, but I wonder if we start exposing linter rules from `@typespec/http` if we have to add an `extends` for it in `typespec-azure-core`.  What do you think @catalinaperalta?

Comment (daviwil) [reply packages/http/src/validate.ts]: Updated the design to use the new linting framework instead. @timotheeguerin, how does it look?  Also, can you answer the question I asked above?

Comment (catalinaperalta) [reply packages/http/src/validate.ts]: @daviwil Yes we'd need to extend this in typespec-azure-core to keep checking for the warnings coming from `@typespec/http`

File: packages/http/src/state.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// FIXME - This is a workaround for the circular dependency issue when loading
```
Comment (timotheeguerin) [inline packages/http/src/state.ts:1]: :( showing up everywhere now

File: packages/http/src/rules/operation-reference-container-route.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { Operation, createRule, paramMessage } from "@typespec/compiler";
+import { getRoutePath } from "../route.js";
+import { OperationContainer } from "../types.js";
+
+export const operationReferenceContainerRouteRule = createRule({
+  name: "operation-reference-container-route",
```
Comment (timotheeguerin) [inline packages/http/src/rules/operation-reference-container-route.ts:6]:
```suggestion
  name: "op-reference-container-route",
```

nit: would be a little nicer to use the short name for those?

**Review by timotheeguerin — APPROVED** [review]
Looks great I think we probably want this to be added to the azure core and arm rulesets too.

File: packages/http/src/rules/operation-reference-container-route.ts (hunk 3 above)
Comment (daviwil) [reply packages/http/src/rules/operation-reference-container-route.ts:6]: Yeah, it's a bit long

File: packages/http/src/validate.ts (hunk 1 above)
Comment (daviwil) [reply packages/http/src/validate.ts]: Thanks!

File: packages/http/src/lib.ts [hunk 4]
```diff
@@ -125,9 +126,16 @@ const libDefinition = {
...
-} as const;
-
-const httpLib = createTypeSpecLibrary(libDefinition);
-const { reportDiagnostic, createDiagnostic, createStateSymbol } = httpLib;
+  linter: {
+    rules: [operationReferenceContainerRouteRule],
+    ruleSets: {
+      all: {
+        enable: {
+          [`@typespec/http/${operationReferenceContainerRouteRule.name}`]: true,
```
Comment (markcowl) [inline packages/http/src/lib.ts:134]: I feel like we need to add this rule to the azure-core ruleset in a separate PR

**Review by markcowl — APPROVED** [review]

File: packages/http/src/lib.ts (hunk 4 above)
Comment (daviwil) [reply packages/http/src/lib.ts]: Sent PR https://github.com/Azure/typespec-azure/pull/3546 to add this

**Pull request merged**